src/storage/src/hummock/local_version_manager.rs

```diff
@@ -511,21 +508,57 @@ impl LocalVersionManager {
         for result in join_all(join_handles).await {
             result.expect("should be able to join the flush handle")
         }
-        let (order_index, task_payload, task_write_batch_size) = match self
-            .local_version
-            .write()
-            .get_mut_shared_buffer(epoch)
-            .and_then(|shared_buffer| shared_buffer.new_upload_task(SyncEpoch))
-        {
-            Some(task) => task,
-            None => {
-                tracing::trace!("sync epoch {} has no more task to do", epoch);
-                return Ok(0);
+        let notify1 = self.sync_epoch_notify.clone();
+        // TODO(xuxinhao): modify it after supporting uploading multiple shared buffers.#4442
+        loop {
+            let epoch1 = self
+                .local_version
+                .read()
+                .iter_shared_buffer()
+                .find(|(key, _)| key > &&self.max_sync_epoch.read().clone())
+                .map(|(key, _)| key)
+                .cloned();
+            if epoch1.is_none() || epoch1.unwrap() == epoch {
+                break;
+            } else {
+                notify1.notified().await;
             }
+        }
+        self.sync_epoch_notify.notify_waiters();
```

Comment:
I am confused about why we need this logic:
1. Looks like it is used to ensure the order of the sync operation. Why do we need that before we introduce multi-epoch upload?
2. We wait for `max_sync_epoch `to update in a loop and do `notify` in L527 but before L527 we never update `max_sync_epoch`... There can be deadlocks.

Reply:
1. Because we need to get the data in descending order by epoch. So we need data in  sync_uncommitted_tasks_ssts to be in order.
2. If we don't sync some epochs in shared buffer, It will be locked until we update max_uncommitted_epoch and update shared buffer.